File: CSE241 - Fall 2019/hw8/src/GTUVector.java
import java.lang.reflect.Array;
import java.security.InvalidParameterException;

public class GTUVector<T> extends GTUContainer<T>{

    public GTUVector(Class<T[]> type){
        type_ = type;
        capacity = 1;
        used = 0;
        data = type.cast(Array.newInstance(type.getComponentType(),capacity));
    }

    public class GTUVectorIterator implements GTUIterator<T>{
        int pos = 0;

        public T next(){
            return data[pos++];
        }

        public boolean hasNext(){
            if(pos < used)
                return true;
            
            return false;
        }
    }
    @Override
    public boolean empty(){
        return (used == 0);
    }
    @Override
    public void clear(){
        used = 0;
        data = null;
    }
    @Override
    public int max_size(){
        return capacity;
    }

    @Override
    public int size(){
        return used;
    }

    @Override
    public void erase(Object obj){
        boolean check = false;
        int index = -1;
        for(int i = 0 ; i < used ; ++i){
            if(data[i] == obj)
                check = true;
                index = i;
        }

        if(check){
            for(int i = index ; i < used - 1; ++i){
                data[i] = data[i + 1];
            }
            --used;

        }
    }

    @Override
    public void insert(T obj) throws InvalidParameterException{
        if(capacity <= used || capacity == 0){
            T[] temp = type_.cast(Array.newInstance(type_.getComponentType(), capacity * 2));
            for(int i = 0 ; i < used ; ++i){
                temp[i] = data[i];         
            }
            
            data = type_.cast(Array.newInstance(type_.getComponentType(), capacity * 2));
            capacity *= 2;
            for(int i = 0 ; i < used ; ++i)
                data[i] = temp[i];
            data[used] = obj;
            ++used;
        } 
        else{
            data[used] = obj;
            ++used;
        }
    }

    @Override
    public boolean contains(Object o){

        for(int i = 0 ; i < used ; ++i){
            if(data[i] == o)
                return true;
        }
        return false;
    }

    public GTUIterator<T> iterator(){
        return new GTUVectorIterator();
    }





}

File: CSE241 - Fall 2019/hw6/GTUVector.cpp
#include "GTUVector.h"


namespace kal{

    template <typename T>
    GTUVector<T>::GTUVector(){
        capacity = 1;
        used = 0;
        std::shared_ptr<T> temp(new T[capacity], std::default_delete<T[]>());
        this->data = temp;
    }

    template <typename T>
    int GTUVector<T>::max_size() const noexcept{
        return capacity;
    }

    template <typename T>
    int GTUVector<T>::size() const noexcept{      
        return used;
    }


    template <typename T>
    bool GTUVector<T>::empty() const noexcept{
        return (used == 0);
    }

    template<typename T>
    void GTUVector<T>::clear() noexcept{
        this->used = 0;
        this->data = nullptr;
    }
    template<typename T>
    int& GTUVector<T>::operator[](int index){
        if(index >= used){
            std::cout << "Index is out of bounds" << endl;
            exit(0);
        }
        return data.get()[index];
    }


    template<typename T>
    void GTUVector<T>::erase(const T& val){
        int count = 0,index = -1,i = 0;
        std::shared_ptr< T > temp(new T[capacity], std::default_delete<T[]>());
        for(auto it = this->begin() ; it != this->end() ; ++it){
            ++count;
            if( (*it) == val){            
                index = count - 1;
            }
            
        }
        if(index == -1) {
         cout << "This value is not in the vector" << endl;
         return;
        }
        if(index == used - 1) --used;
        else{
            int j = index;
            for(auto it = begin() + index ; it != this->end() ; ++it){
                data.get()[j] = data.get()[j + 1];
                ++j;
            }
            --used;
        }

    }
    template<typename T>
    void GTUVector<T>::erase(GTUIterator<T> position){
        int index = -1,count = 0;
        for(auto it = this->begin() ; it != this->end() ; ++it){
            ++count;
            if( it == position){            
                index = count - 1;
            }

        }
        if(index == -1){
            cout << "Invalid iterator" << endl;
            return;
        }
        if(index == used - 1) --used;
        else{
            int j = index;    
            for(auto it = position  ; it != this->end() ; ++it){
                data.get()[j] = data.get()[j + 1];
                ++j;
            }
            --used;
        }
    }

    template <typename T>
    void GTUVector<T>::insert(const T& val){
        int i = 0;
        if(capacity <= used || capacity == 0){
            std::shared_ptr<T> temp(new T[capacity * 2], std::default_delete<T[]>());
            for(auto it = this->begin() ; it != this->end() ; ++it){
                temp.get()[i] = *(it);
                ++i;
            }
            temp.get()[i] = val;
             this->data = nullptr;
             this->data = temp;           
             capacity = capacity * 2;
             ++used;
        }
        else{       
            data.get()[used] = val;
            ++used;              
        }

        
    }


}